hassil/parse_expression.py

```diff
@@ -0,0 +1,189 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from .expression import (
+    Expression,
+    ListReference,
+    RuleReference,
+    Sentence,
+    Sequence,
+    SequenceType,
+    TextChunk,
+)
+from .parser import (
+    GROUP_END,
+    GROUP_START,
+    LIST_END,
+    LIST_START,
+    OPT_END,
+    OPT_START,
+    RULE_END,
+    RULE_START,
+    ParseChunk,
+    ParseError,
+    ParseType,
+    next_chunk,
+    remove_delimiters,
+)
+from .util import normalize_text
+
+
+@dataclass
+class ParseMetadata:
+    """Debug metadata for more helpful parsing errors."""
+
+    file_name: str
+    line_number: int
+    intent_name: Optional[str] = None
+
+
+class ParseExpressionError(ParseError):
+    def __init__(self, chunk: ParseChunk, metadata: Optional[ParseMetadata] = None):
+        super().__init__()
+        self.chunk = chunk
+        self.metadata = metadata
+
+    def __str__(self) -> str:
+        return f"Error in chunk {self.chunk} at {self.metadata}"
+
+
+def ensure_alternative(seq: Sequence):
+    if seq.type != SequenceType.ALTERNATIVE:
+        seq.type = SequenceType.ALTERNATIVE
+
+        # Collapse items into a single group
+        seq.items = [
+            Sequence(
+                type=SequenceType.GROUP,
+                items=seq.items,
+            )
+        ]
+
+
+def parse_group_or_alt(
+    seq_chunk: ParseChunk, metadata: Optional[ParseMetadata] = None
+) -> Sequence:
+    seq = Sequence(type=SequenceType.GROUP)
+    if seq_chunk.parse_type == ParseType.GROUP:
+        seq_text = remove_delimiters(seq_chunk.text, GROUP_START, GROUP_END)
+    elif seq_chunk.parse_type == ParseType.OPT:
+        seq_text = remove_delimiters(seq_chunk.text, OPT_START, OPT_END)
+    else:
+        raise ParseExpressionError(seq_chunk, metadata=metadata)
+
+    item_chunk = next_chunk(seq_text)
+    last_seq_text = seq_text
+
+    while item_chunk is not None:
+        if item_chunk.parse_type in {
```

Comment:
I once read but can't find it that tuples are faster because easier to optimize

Reply:
Ah, good idea. I did this with the delimiter groups too.